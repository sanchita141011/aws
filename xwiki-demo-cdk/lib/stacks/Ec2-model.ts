import { StackProps } from '@aws-cdk/core'

/**
 * An insterface class for extending the props of EC2XwikiDemo class
 * 
 * <p> This is used for extending the properties of EC2XwikiDemo class
 * here xwiki is added as the proprties that will be used to provide download 
 * link of the xwiki to EC2XwikiDemo Class from the config file.
 */

export interface ec2props extends StackProps{

    xwiki: string

}
