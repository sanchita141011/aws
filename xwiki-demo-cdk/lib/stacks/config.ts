export const xwikidownload = 'https://nexus.xwiki.org/nexus/content/groups/public/org/xwiki/platform/xwiki-platform-distribution-flavor-jetty-hsqldb/13.1/xwiki-platform-distribution-flavor-jetty-hsqldb-13.1.zip'
export const region ="us-east-1"
