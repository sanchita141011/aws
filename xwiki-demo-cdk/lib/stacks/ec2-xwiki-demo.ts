#!/usr/bin/env node
/*
 * See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */
import * as cdk from '@aws-cdk/core'
import {Vpc,SecurityGroup, Peer,Port, Instance, InstanceType, InstanceClass, InstanceSize, MachineImage, AmazonLinuxGeneration, SubnetType} from '@aws-cdk/aws-ec2'
import * as iam from '@aws-cdk/aws-iam'
import * as fs from 'fs'

export class EC2XwikiDemo extends cdk.Stack {
  constructor (scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props)

    const defaultVpc = new Vpc(this, 'VPC', { 
      cidr: "10.0.0.0/16",
      maxAzs:2,
      subnetConfiguration: [
        {
          name: 'public',
          subnetType: SubnetType.PUBLIC,
          cidrMask: 27
        }
      ]
     })

    const role = new iam.Role(
      this,
      'ec2-xwiki-demo-instance-role',
      { assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com') }
    )

    const securityGroup = new SecurityGroup(
      this,
      'ec2-xwiki-demo-sg',
      {
        vpc: defaultVpc,
        allowAllOutbound: true, // will let xwiki instance send outbound traffic
        securityGroupName: 'ec2-xwiki-demo-sg'
      }
    )

    // configuring the security group to allow inbound traffic at specific ports
    securityGroup.addIngressRule(
      Peer.anyIpv4(),
      Port.tcp(22),
      'Allows SSH access from Internet'
    )

    securityGroup.addIngressRule(
      Peer.anyIpv4(),
      Port.tcp(80),
      'Allows HTTP access from Internet'
    )

    securityGroup.addIngressRule(
      Peer.anyIpv4(),
      Port.tcp(443),
      'Allows HTTPS access from Internet'
    )

    securityGroup.addIngressRule(
      Peer.anyIpv6(),
      Port.tcp(8080),
      'Allows port 8080 on Ipv6'
    )

    securityGroup.addIngressRule(
      Peer.anyIpv4(),
      Port.tcp(8080),
      'Allows port 8080 on Ipv4'

    )

    // provesioning the EC2 instance inside the VPC containg xwiki
    const instance = new Instance(this, 'ec2-xwiki-demo', {
      vpc: defaultVpc,
      role: role,
      securityGroup: securityGroup,
      instanceName: 'ec2-xwiki-demo',
      instanceType: InstanceType.of( // t2.micro has free tier usage in aws
        InstanceClass.T2,
        InstanceSize.MEDIUM
      ),
      machineImage: MachineImage.latestAmazonLinux({
        generation: AmazonLinuxGeneration.AMAZON_LINUX_2
      }),

      keyName: 'test'// user will first create this key in console

    })
    // adding the commands that will install xwiki in the instance
    instance.addUserData(
      fs.readFileSync('lib/stacks/user_script.sh', 'utf8')
    )

    // cdk lets us output prperties of the resources we create after they are created
    // we want the ip address of this new instance so we can ssh into it later
    new cdk.CfnOutput(this, 'xwiki-demo-instance-output', {
      value: instance.instancePublicIp
    })
  }
}
