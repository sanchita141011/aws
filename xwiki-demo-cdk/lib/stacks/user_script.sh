#! /bin/bash
# become root user

sudo yum -q -y install java-1.8.0-openjdk
sudo yum -q -y install unzip
mkdir xwikihome
wget -q -O xwiki_packer.zip https://nexus.xwiki.org/nexus/content/groups/public/org/xwiki/platform/xwiki-platform-distribution-flavor-jetty-hsqldb/13.1/xwiki-platform-distribution-flavor-jetty-hsqldb-13.1.zip
unzip -q xwiki_packer.zip -d /home/ec2-user/xwikihome



