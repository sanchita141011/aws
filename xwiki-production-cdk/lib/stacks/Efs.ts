import * as cdk from '@aws-cdk/core';
import { FileSystem, PerformanceMode } from '@aws-cdk/aws-efs';
import {SecurityGroup, SubnetType} from  '@aws-cdk/aws-ec2';
import { efsprops } from '../models/efs-model'


export class XwikiEfs extends cdk.Stack{

  public readonly xwikiEfs: FileSystem;
    constructor(scope: cdk.App, id: string, props: efsprops){
        super(scope, id, props)

        

        const xwikiEfsSg = new SecurityGroup(this, 'trcXWikiEfsSecurityGroup', {
            vpc: props.vpc,
            allowAllOutbound: true,
            description: `Security Group for XWiki EFS`
          });
      
          const xwikiEfs = new FileSystem(this, 'trcXWikiFileSystem', {
            vpc: props.vpc,
            enableAutomaticBackups: true,
            encrypted: true,
            kmsKey: props.encryptionkey,
            performanceMode: PerformanceMode.GENERAL_PURPOSE,
            securityGroup: xwikiEfsSg,
            vpcSubnets: props.vpc.selectSubnets(
              {
                subnetType: SubnetType.PRIVATE
              }
            )
          });

          this.xwikiEfs=xwikiEfs;
      
    }
}