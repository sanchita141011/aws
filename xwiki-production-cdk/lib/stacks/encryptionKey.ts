import { Key } from '@aws-cdk/aws-kms'
import * as cdk from '@aws-cdk/core'

export class XwikiEncryptionKey extends cdk.Stack {
    public readonly EncryptionKey: Key;
    public readonly SecretEncryptionKey: Key;
    constructor (scope: cdk.App, id: string, props?: cdk.StackProps) {
      super(scope, id, props)

      const EncryptionKey = new Key(this, 'xwiki_encryption_key', {
        alias: 'xwiki-key',
        description: 'Encryption Key for XWiki Storage Resources',
        enableKeyRotation: true,
        enabled: true,
        trustAccountIdentities: true
      })

      const SecretEncryptionKey = new Key(this, 'xwiki_secret_encryption key', {
        alias: 'xwiki-secret-key',
        description: 'Encryption Key for XWiki Secrets',
        enableKeyRotation: true,
        enabled: true,
        trustAccountIdentities: true
      })

      this.EncryptionKey = EncryptionKey
      this.SecretEncryptionKey = SecretEncryptionKey
    }
}
