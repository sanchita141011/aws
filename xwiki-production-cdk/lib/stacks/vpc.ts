#!/usr/bin/env node
/*
 * See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

import { DefaultInstanceTenancy, NatProvider, SubnetType, Vpc } from '@aws-cdk/aws-ec2'
import * as cdk from '@aws-cdk/core'

/**
 * The class XWikiVpx create a stack to provision a new VPC in user's account
 * The VPC will constain 4 subnets, 2 private and 2 public.
 */
export class XWikiVpc extends cdk.Stack {
    public readonly xwikivpc: Vpc;
    constructor (scope: cdk.App, id: string, props?: cdk.StackProps) {
      super(scope, id, props)

      this.xwikivpc = new Vpc(this, 'xwiki-vpc', {
        cidr: '10.42.42.0/24', //this is enough for this installation but not if you want to deploy other services inside this.
        defaultInstanceTenancy: DefaultInstanceTenancy.DEFAULT,
        maxAzs: 2,
        natGatewayProvider: NatProvider.gateway(),
        natGateways: 1,
        subnetConfiguration: [
          {
            name: 'public',
            subnetType: SubnetType.PUBLIC,
            cidrMask: 27
          },
          {
            name: 'private-database',
            subnetType: SubnetType.PRIVATE,
            cidrMask: 26
          }
        ]
      })
    }
}
