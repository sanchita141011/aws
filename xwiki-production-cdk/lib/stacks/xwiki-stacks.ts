
import { Stack, App, Duration, CfnOutput } from '@aws-cdk/core'
import { AwsLogDriver, Cluster, ContainerImage, FargatePlatformVersion, FargateService, FargateTaskDefinition, Secret as EcsSecret } from '@aws-cdk/aws-ecs'
import { AuroraCapacityUnit, AuroraMysqlEngineVersion, DatabaseClusterEngine, ServerlessCluster, SubnetGroup } from '@aws-cdk/aws-rds'
import { FileSystem, PerformanceMode } from '@aws-cdk/aws-efs'
import { Peer, Port, SecurityGroup, SubnetType } from '@aws-cdk/aws-ec2'
import { Key } from '@aws-cdk/aws-kms'
import { LogGroup, RetentionDays } from '@aws-cdk/aws-logs'
import { Role, ServicePrincipal } from '@aws-cdk/aws-iam'
import { Secret } from '@aws-cdk/aws-secretsmanager'
import { ApplicationLoadBalancer, ApplicationProtocol } from '@aws-cdk/aws-elasticloadbalancingv2'
import { xwikistackprops } from '../models/xwiki-stack-models'
import { xwikiVersion } from '../config'

/**
 * The XwikiProductionStacks is used for confuguring stacks in the vpc provissioned by class XwikiVpc.
 * 
 * <p> this stack will provission two encryption keys using AWS KMS, ECS fargate clusture, RDS instance
 * with MySQL and a loadbalncer. The output after running the installation will be DNS Endpoint for connecting to the XWiki Installation
 * The target for loadbalancer is Fargate Service, which is configured in the private subnet part of the VPC
 * The security group used for Fargate sevice will allow inbound traffic connection to HTTP port, and will allow connection to DB for xwiki
 * The fargate service uses task defination that will add container and use image from dockerhub. 
 * This class will provission RDS with mysql for XWiki DB using serverless clusture with Aurora MySQL engine. 
 * The password used by xwikirds is autogenertatd of length 16. This RDS resides in private subnet part of the VPC.
 * 
 */
export class XWikiProductionStacks extends Stack {
    constructor (scope: App, id: string, props: xwikistackprops) {
      super(scope, id, props)

      const xwikiEncryptionKey = new Key(this, 'XWikiEncryptionKey', { //encryption key to be used by the file system and rds
        alias: `xwiki`,
        description: `Encryption Key for XWiki Storage Resources`,
        enableKeyRotation: true,
        enabled: true,
        trustAccountIdentities: true,
      });
  
      const xwikiSecretEncryptionKey = new Key(this, 'XWikiSecretEncryptionKey', { //used for fenerating password for rds
        alias: `xwiki-secret`,
        description: `Encryption Key for XWiki Secrets`,
        enableKeyRotation: true,
        enabled: true,
        trustAccountIdentities: true,
      });
  
      const xwikiEfsSg = new SecurityGroup(this, 'XWikiEfsSecurityGroup', { //security group for file system.
        vpc: props.vpc,
        allowAllOutbound: true,
        description: `Security Group for XWiki EFS`
      });
  
      const xwikiEfs = new FileSystem(this, 'XWikiFileSystem', { // File System that will conatin static xwiki files
        vpc: props.vpc,
        enableAutomaticBackups: true,
        encrypted: true,
        kmsKey: xwikiEncryptionKey,
        performanceMode: PerformanceMode.GENERAL_PURPOSE,
        securityGroup: xwikiEfsSg,
        vpcSubnets: props.vpc.selectSubnets(
          {
            subnetType: SubnetType.PRIVATE
          }
        )
      });
  
      const xwikiRdsSg = new SecurityGroup(this, 'XWikiRdsSecurityGroup', { //security group for xwiki RDS
        vpc: props.vpc,
        allowAllOutbound: true,
        description: `Security Group for XWiki RDS`
      });
  
      const xwikiRdsDbSubnetGroup = new SubnetGroup(this, 'XWikiDbSubnetGroup', { //subnet group for RDS
        description: `DB SubnetGroup for XWiki RDS`,
        vpc: props.vpc,
        vpcSubnets: props.vpc.selectSubnets(
          {
            subnetType: SubnetType.PRIVATE
          }
        )
      });
  
      const xwikiRdsPwSecret = new Secret(this, 'XWikiEcsUserPassword', { //Auto-generated password for RDS
        description: `RDS UserSecret for XWiki RDS`,
        encryptionKey: xwikiSecretEncryptionKey,
        generateSecretString: {
          excludePunctuation: true,
          passwordLength: 16
        }
      });
  
      const xwikiRds = new ServerlessCluster(this, 'XWikiDbCluster', { //configuration of RDS for xwiki
        engine: DatabaseClusterEngine.auroraMysql({
          version: AuroraMysqlEngineVersion.VER_2_07_1
        }),
        vpc: props.vpc,
        vpcSubnets: props.vpc.selectSubnets(
          {
            subnetType: SubnetType.PRIVATE
          }
        ),
        credentials: {
          username: 'xwikimysql',
          password: xwikiRdsPwSecret.secretValue,
        },
        backupRetention: Duration.days(7),
        scaling: {
          autoPause: Duration.minutes(0), // AutoPause Disabled
          minCapacity: AuroraCapacityUnit.ACU_1,
          maxCapacity: AuroraCapacityUnit.ACU_8
        },
        securityGroups: [
          xwikiRdsSg
        ],
        defaultDatabaseName: 'xwiki',
        storageEncryptionKey: xwikiEncryptionKey,
        subnetGroup: xwikiRdsDbSubnetGroup
      });
  
      const xwikiFargateCluster = new Cluster(this, 'XWikiCluster', {
        containerInsights: true,
        vpc: props.vpc
      });
  
      const xwikiTaskIamRole = new Role(this, 'XwikiTaskRole', { //IAM role for ECS fargate
        assumedBy: new ServicePrincipal('ecs-tasks.amazonaws.com'),
        description: `IAM Task Role for XWiki ECS Fargate`,
      });
  
      const xwikiTaskDefinition = new FargateTaskDefinition(this, 'XWikiTaskDefinition', { //taskdefination for ecs service
        cpu: 2048,
        memoryLimitMiB: 4096,
        volumes: [
          {
            name: 'EfsPersistendVolume',
            efsVolumeConfiguration: {
              fileSystemId: xwikiEfs.fileSystemId,
              rootDirectory: '/',
              transitEncryption: 'ENABLED'
            }
          }
        ],
        executionRole: xwikiTaskIamRole,
        taskRole: xwikiTaskIamRole
      });
  
      const xwikiLogGroup = new LogGroup(this, 'XWikiLogGroup', {
        retention: RetentionDays.ONE_MONTH
      });
  
      xwikiLogGroup.grantWrite(xwikiTaskIamRole);
  
      const xwikiContainer = xwikiTaskDefinition.addContainer('XWikiImage', { //adding container configuration in taskdefination
        image: ContainerImage.fromRegistry(xwikiVersion),
        environment: {
          'DB_HOST': xwikiRds.clusterEndpoint.hostname,
          'DB_DATABASE': 'xwiki',
          'DB_USER': 'xwikimysql'
        },
        logging: new AwsLogDriver({
          logGroup: xwikiLogGroup,
          streamPrefix: `xwiki`
        }),
        secrets: {
          'DB_PASSWORD': EcsSecret.fromSecretsManager(xwikiRdsPwSecret)
        },
        essential: true,
      });
  
      xwikiContainer.addPortMappings({
        containerPort: 8080
      });
  
      xwikiContainer.addMountPoints({
        containerPath: '/usr/local/xwiki/data',
        readOnly: false,
        sourceVolume: 'EfsPersistendVolume'
      });
  
      const xwikiServiceSecurityGroup = new SecurityGroup(this, 'XWikiTaskSecurityGroup', {
        vpc: props.vpc,
        allowAllOutbound: true,
        description: `SecurityGroup for XWiki ECS Fargate Service`
      });
  
      const xwikiLoadBalancerSecurityGroup = new SecurityGroup(this, 'XWikiAlbSecurityGroup', {
        vpc: props.vpc,
        allowAllOutbound: true,
        description: `SecurityGroup for XWiki Application LoadBalancer`
      });
  
      xwikiLoadBalancerSecurityGroup.addIngressRule(
        Peer.anyIpv4(),
        Port.tcp(80),
        `Allow HTTP Connections for the World to Application LoadBalancer`
      );
  
      const xwikiLoadBalancer = new ApplicationLoadBalancer(this, 'XWikiLoadBalancer', {
        vpc: props.vpc,
        internetFacing: true,
        securityGroup: xwikiLoadBalancerSecurityGroup
      });
  
      const xwikiLoadBalancerListener = xwikiLoadBalancer.addListener('XWikiLoadBalancerHttpListener', {
        protocol: ApplicationProtocol.HTTP,
      });
  
      xwikiServiceSecurityGroup.addIngressRule(
        xwikiLoadBalancerSecurityGroup,
        Port.tcp(8080),
        `Allow HTTP Connections for XWiki ECS Application LoadBalancer`
      );
  
      xwikiEfsSg.addIngressRule(
        xwikiServiceSecurityGroup,
        Port.tcp(2049),
        `Allow NFS Connection for XWiki Service`
      );
  
      xwikiRdsSg.addIngressRule(
        xwikiServiceSecurityGroup,
        Port.tcp(xwikiRds.clusterEndpoint.port),
        `Allow DB Connection for XWiki Service`
      );
  
      const xwikiEcsService = new FargateService(this, 'XWikiService', { //confuguration of fargate service
        cluster: xwikiFargateCluster,
        taskDefinition: xwikiTaskDefinition,
        desiredCount: 1,
        platformVersion: FargatePlatformVersion.VERSION1_4,
        vpcSubnets: props.vpc.selectSubnets(
          {
            subnetType: SubnetType.PRIVATE
          }
        ),
        securityGroups: [
          xwikiServiceSecurityGroup
        ]
      });
  
      xwikiLoadBalancerListener.addTargets('XWikiTargets', { //adding target as ecs service to loadbalancer
        deregistrationDelay: Duration.minutes(1),
        protocol: ApplicationProtocol.HTTP,
        targets: [
          xwikiEcsService
        ],
        healthCheck: {
          healthyHttpCodes: '200,301,302'
        }
      });
  
      new CfnOutput(this, 'XWikiLoadBalancerDns', { //this will output the DNS endpoint to connect to xwiki instance
        value: xwikiLoadBalancer.loadBalancerDnsName,
        description: `DNS Endpoint for connecting to the XWiki Installation`
      });
    }
  }

