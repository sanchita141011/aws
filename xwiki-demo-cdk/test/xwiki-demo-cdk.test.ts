/*
 * See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */
import { expect as expectCDK, haveResourceLike, countResources } from '@aws-cdk/assert'
import * as cdk from '@aws-cdk/core'
import '@aws-cdk/assert/jest'
import { EC2XwikiDemo } from '../lib/stacks/ec2-xwiki-demo'

test('Check InstanceType and SSH KeyName', () => {
  const app = new cdk.App()
  // WHEN
  // const stack= new cdk.Stack(app, 'xwiki-cdk-demo-test', )
  const stack = new EC2XwikiDemo(app, 'xwiki-cdk-demo', {
    env: {
      account: '656019072197',
      region: 'us-east-1'
    }
  })

  // expect(stack).toBe(countResources("AWS::EC2::Instance", 1));
  // THEN
  expectCDK(stack).to(
    countResources('AWS::EC2::Instance', 1))
    
  expectCDK(stack).to(
    haveResourceLike('AWS::EC2::Instance', {
      InstanceType: 't2.medium',
      KeyName: 'test'

    })
  )
})
