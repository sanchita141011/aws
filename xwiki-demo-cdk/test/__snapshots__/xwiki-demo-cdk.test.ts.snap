// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`demo stack matches the snapshot 1`] = `Object {}`;
